docs/source/gbd2019_models/hemoglobin_anemia_and_iron_deficiency/hemoglobin/index.rst

```diff
@@ -6,4 +6,290 @@ Hemoglobin Model
++++++++++++++++++++++++++++
+
+.. todo::
+
+  List assumptions and limitations
+
+Data Description Tables
++++++++++++++++++++++++
+
+.. list-table:: Distribution Parameters
+  :widths: 15, 30, 10
+  :header-rows: 1
+
+  * - Parameter
+    - Value
+    - Note
+  * - Hemoglobin mean
+    - MEID 10487
+    - source='epi', decom_step='step4'
+  * - Hemoglobin standard deviation
+    - MEID 10488
+    - source='epi', decomp_step='step4'
+  * - XMAX
+    - 220
+    - 
+  * - Euler's constant
+    - 0.57721566490153286060651209008240243104215933593992
+    - 
+  * - Gamma distribution weight
+    - 0.4
+    - 
+  * - Mirror Gumbel distribution weight
+    - 0.6
+    - 
+
+Below is R code written to randomly sample hemoglobin concentration values from the hemoglobin distribution parameters and constants defined in the tables above. Additionally, the code block contains functions that will evaluate the proportion of the distribution below a given threshold. This code was adapted from the GBD stash code found `here <https://stash.ihme.washington.edu/projects/MNCH/repos/anemia/browse/model/envelope>`_, specifically the *DistList_mv2p.R* and *fit_ensemblemv2p_parallel.R* files. 
+
+.. note::
+
+  Attempts to move this code to python have not been successful in validating to GBD while using the R code has been successful. The recommended approach is to use the python package `rpy2` to run this R code within python.
+
+.. code-block:: R
+
+  # define constants
+  XMAX = 220
+  EULERS_CONSTANT = 0.57721566490153286060651209008240243104215933593992
+  w = c(0.4,0.6)
+
+  # import standard R functions for the gamma distributions (pgamma and rgamma)
+  pacman::p_load(data.table,actuar)
+
+  # function to calculate gamma distribution parameters from mean and variance
+  gamma_mv2p = function(mn, vr){
+    list(shape = mn^2/vr,rate = mn/vr)}
+
+  # function to calculate mirror gumbel distribution parameters from mean and variance 
+  mgumbel_mv2p = function(mn, vr){
+    list(
+      alpha = XMAX - mn - EULERS_CONSTANT*sqrt(vr)*sqrt(6)/pi,
+      scale = sqrt(vr)*sqrt(6)/pi)}
+  # function to randomly sample n times from mirror gumbel distribution
+  rmgumbel = function(n, alpha, scale){
+    mn = alpha + scale*EULERS_CONSTANT
+    rgumbel(n, alpha+XMAX-(2*mn), scale)}
+
+  # function to sample from ensemble hemoglobin distribution using functions defined above
+    # n = number of samples to take
+    # mn = mean hemoglobin concentration
+    # vr = hemoglobin distribution variance (standard deviation squared)
+    # w = list of ensemble distribution weights c(gamma_weight, mirror_gumbel_weight)
+  sample_from_hemoglobin_distribution <- function(n, mn, vr, w){
+
+    ##parameters
+    params_gamma = gamma_mv2p(mn, vr)
+    params_mgumbel = mgumbel_mv2p(mn, vr)
+
+    ##weighting
+    sample_list = sum(
+      w[1] * rgamma(n, params_gamma$shape, params_gamma$rate), 
+      w[2] * rmgumbel(n, params_mgumbel$alpha, params_mgumbel$scale)
+    )
+    sample_list
+  }
```

Comment:
This is not actually the correct way to sample from an ensemble distribution. However, this *is* what Vivarium currently does, and it seems to give results that are fairly close to the right answer (except in the tails of the distribution).

I am in the process of writing up some documentation to describe how to correctly sample from ensemble distributions using propensities. Basically, the algorithm in this case would be:
1. Choose Gamma with probability 0.4 or mirror Gumbel with probability 0.6.
2. Sample from the chosen distribution.

In other words, each draw from the ensemble distribution samples from *either* Gamma or mirror Gumbel, but not both.

I'm not sure how to write the correct R code to do this. Until we have corrected sampling code, I think we should remove the incorrect version to avoid confusion.

Reply:
It looks like in the new version, all of the gamma draws will be placed before all of the mirror Gumbel draws - is that correct, or am I misunderstanding the R code?

If that's the case, I think there needs to be an extra step shuffling all the draws together so that we get an independent sequence of draws from the mixture.

Reply:
Also, if n=1, I think this updated method will always sample from mirror Gumbel, which is incorrect. So I think we need a different strategy for generating the draws.

Reply:
I should have said that **on average**, 40% of the draws will come from the Gamma distribution, and 60% of the draws will come from the mirror Gumbel distribution. The actual number from each won't exactly equal this ratio -- it will be random with a binomial distribution.

So one solution would be to sample the number of gamma draws as n1 = Binomial(n, 0.4), then take the number of mirror Gumbel draws to be n - n1, and then you'd have to randomly shuffle the draws to make sure they were in no particular order.

Or, a simple non-vectorized solution would be to just put the two steps I originally described above into a for loop of length n.

Reply:
Ok, looks good